### ISSUE TITLE
avoid_types_as_parameter_names should trigger on catch functions

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#1493]
Right now a user can write the following, which is not caught by [avoid_types_as_parameter_names](https://github.com/dart-lang/linter/blob/master/test/rules/avoid_types_as_parameter_names.dart):

```dart
try {
  // code;
} catch (ArgumentError) {
  // code;
}
```

but I think it should be caught.

**Comment by srawlins** [comment]
I'd  super duper prefer that this be caught by avoid_types_as_parameter_names, rather than a new rule,  since it is conceptually one rule. AFAIK the only thing preventing that might be whether the Flutter repo has a lot of pre-existing violations, but I can't imagine it does.

**Comment by bwilkerson** [comment]
I agree that it would be good to detect this.

It does seem a little different in that everywhere else that this lint applies to it is possible to write both a type name and a parameter name and the type name is optional, but it isn't valid to write a type name in this location. But I do see the similarities, and I'm not sure I have a strong opinion.

Does the lint rule currently check for uses of the parameter name within the body of the function (and assume that the hiding was intentional in those cases)?

**Comment by natebosch** [comment]
I strongly agree that we should catch this `catch (SomeType)` is a common mistake when the author intended `on SomeType` or `on SomeType catch (_)`.

**Comment by grouma** [comment]
+1 Just ran into this issue. I always forget the syntax but I know there is a way to catch only specific types. I naturally try `catch (SomeType)` which is obviously wrong but it "works" without warning.

**Comment by pq** [comment]
Thanks for chiming in.  Taking a look now! 👍

**Comment by pq** [comment]
https://github.com/dart-lang/linter/pull/2060
